lazyflow/distributed/TaskOrchestrator.py

```diff
@@ -0,0 +1,103 @@
+from mpi4py import MPI
+import enum
+from typing import Iterable, TypeVar, Generic, Callable, Tuple
+import uuid
+import logging
+
+print(f"THIS IS MY NAME: {__name__}")
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+# message payload to signal a worker that it should terminate. Value is arbitrary but should be universally unique
+COMMAND_STOP_WORKER = "COMMAND_STOP_WORKER-eb23ae13-709e-4ac3-931d-99ab059ef0c2"
+UNIT_OF_WORK = TypeVar("UNIT_OF_WORK")
+
+
+@enum.unique
+class Tags(enum.IntEnum):
+    """Tags are arbitrary ints used to identify the type/purpose of a message in MPI"""
+
+    TASK_DONE = 1  # workers send messages tagged with TASK_DONE when they finished processing a unit of work
+    WORK = enum.auto()  # units of work are tagged with "WORK" and sent to workers for processing
+
+
+class _Worker(Generic[UNIT_OF_WORK]):
+    """A representation of a remote worker"""
+
+    def __init__(self, comm, rank: int):
+        self.comm = comm  # MPI communication channel
+        self.rank = rank  # mpi worker rank, analogous to a worker ID
+        self.stopped = False
+
+    def send(self, unit_of_work: UNIT_OF_WORK):
+        logger.debug(f"Sending unit_of_work {unit_of_work} to worker {self.rank}...")
+        self.comm.send(unit_of_work, dest=self.rank, tag=Tags.WORK)
+
+    def stop(self):
+        self.send(COMMAND_STOP_WORKER)
+        self.stopped = True
+
+
+class TaskOrchestrator(Generic[UNIT_OF_WORK]):
+    """Coordinates work amongst MPI processes.
+
+    In order to use this class, applications must be launched with mpirun: e.g.: mpirun -N <num_workers> ilastik.py
+    """
+
+    def __init__(self, comm=None):
+        self.comm = comm or MPI.COMM_WORLD
+        self.rank = self.comm.Get_rank()
+        num_workers = self.comm.size - 1
+        if num_workers <= 0:
+            raise Exception("Trying to orchestrate tasks with no workers!!!")
```

Comment:
could be maybe a `RuntimeError`? Just so that noone has to do bare `catch Exception` when running this. Or even better have your own `OrchestratorException` class for users...

Reply:
Indeed. Changed to a `ValueError`, which seems more appropriate than an `Exception`